cpp/src/arrow/compute/kernels/cast.cc

```diff
@@ -290,7 +392,8 @@ struct CastFunctor<O, I, typename std::enable_if<is_float_truncate<O, I>::value>
                                                input.length);
         for (int64_t i = 0; i < input.length; ++i) {
           auto out_value = static_cast<out_type>(*in_data);
-          if (ARROW_PREDICT_FALSE(static_cast<in_type>(out_value) != *in_data)) {
+          if (ARROW_PREDICT_FALSE(is_valid_reader.IsSet() &&
```

Comment:
This additional check isn't tested, per above, per above

Reply:
I don't follow. The way I read it, it borks on "unset" invalid data.

Reply:
Right, but in the test case in this patch, `IsSet()` is always true. If you removed this change, then the test case would still pass. So I'm proposing to add a test case where there's a bad value in a null slot